App.js

```diff
@@ -0,0 +1,89 @@
+import TodoInput from "./components/TodoInput.js";
+import TodoList from "./components/TodoList.js";
+import TodoCount from "./components/TodoCount.js";
+import TodoFilter from "./components/TodoFilter.js";
+
+function App() {
+  this.data = JSON.parse(localStorage.getItem("todos")) || [];
+  this.app = this; // 없애볼까..
+
+  // 컴포넌트 생성
+  const todoInput = new TodoInput(this.app);
+  const todoList = new TodoList(this.app, this.data);
+  const todoCount = new TodoCount(this.app, this.data);
+  const todoFilter = new TodoFilter(this.app);
+
+  this.newTodo = (text) => {
+    this.data.push({
+      text,
+      id: Date.now(), // 숫자가 의미하는 바는?
+      completed: false,
+    });
+    this.setState(this.data);
+  };
```

Comment:
48번째 줄에서 보면 석진님께서 `원본 데이터`를 신경 쓰고 계신것 같더라구요.

저도 그렇게 알고 있습니다. 원본 데이터를 변경시키지 않는 선에서 상태를 업데이트 쳐야한다고 알고 있어요. :)

그렇다면 이부분은 아마.. 거기에 해당하지 않는 것같아요.

바로 `this.data` 에 바로 `push` api를 통해 집어 넣으니까, 원본 데이터를 수정하고 있는 것 아닐까 생각이 듭니다.

```suggestion
  this.newTodo = (text) => {
    const newData = this.data;
    newData = [
      ...this.data,
      {
      text,
      id: Date.now(), // 숫자가 의미하는 바는?
      completed: false,
      }
    ];
    this.setState(newData);
  };
```

이렇게 하면 `this.data` 원본을 유지하면서도, 새로운 `todo` 가 입력되는 내용을 반영 해서 `setState` 가 될것 같아요.

Reply:
오 Three dots 표기법 개념이 확실치 않아서 도입 못했는데 이렇게 쓰는구나.. 근데 훨씬 좋게 개선됐네요 굿굿